scss/ScssParser.g4

```diff
@@ -31,393 +32,523 @@ parser grammar ScssParser;
+    : ( identifier | Times )? Pipe
+    ;
 
-conditions
-  : condition (COMBINE_COMPARE conditions)?
-  | NULL_
-  ;
+universal
+    : typeNamespacePrefix? Times
+    ;
 
-condition
-  : commandStatement (( EQEQ | LT | GT | NOTEQ) conditions)?
-  | LPAREN conditions RPAREN
-  ;
+className
+    : Dot ( Minus | identifier | interpolation )+
+    ;
 
+interpolation
+    : namespace_? Hash BlockStart namespace_? ( ifExpression | value | parentRef ) BlockEnd measurment?
+    ;
 
-variableDeclaration
-  : variableName COLON (propertyValue | listBracketed | map_) POUND_DEFAULT? SEMI
-  ;
+parentRef
+    : Amp
+    ;
 
+attrib
+    : Lbrack typeNamespacePrefix? identifier
+    ( ( PrefixMatch | SuffixMatch | SubstringMatch | Eq | Includes | DashMatch )
+    ( identifier | String_ ) )? Rbrack
+    ;
 
-//for
-forDeclaration
-  : AT_FOR variableName FROM fromNumber (TO|THROUGH) throughNumber block
-  ;
+pseudo
+    : Colon Colon? ( interpolation | identifier | functionalPseudo )
+    ;
 
-fromNumber
-  : Number
-  ;
+functionalPseudo
+    : Ident Lparen pseudoParameter+ Rparen
+    ;
 
-throughNumber
-  : Number
-  | functionCall
-  ;
+pseudoParameter
+    : ( ( value | className | interpolation ) Comma?)
+    ;
 
-//while
-whileDeclaration
-  : AT_WHILE conditions block
-  ;
+negation
+    : PseudoNot negationArg Rparen
+    ;
 
-//EACH
-eachDeclaration
-  : AT_EACH variableName (COMMA variableName)* IN eachValueList block
-  ;
+negationArg
+    : typeSelector
+    | universal
+    | Hash
+    | className
+    | attrib
+    | pseudo
+    ;
 
-eachValueList
-  : commandStatement
-  | list_
-  | map_
-  ;
+// Operators
+operator_
+    : Div
+    | Times
+    | Minus
+    | Plus
+    | Greater
+    | Less
+    | Greater Eq
+    | Less Eq
+    | Eq Eq?
+    | NotEq
+    | And
+    | Or
+    | Not
+    ;
 
-//Imports
-importDeclaration
-  : IMPORT referenceUrl SEMI
-  | REQUIRE  referenceUrl SEMI?
-  | USE referenceUrl asClause? withClause? SEMI?
-  | FORWARD referenceUrl asClause? (showClause | hideClause)?
-  ;
+value
+    : unit
+    | number
+    | boolean
+    | calc
+    | rotate
+    | rgba
+    | var_
+    | uri
+    | repeat
+    | Format
+    | String_
+    | functionCall
+    | variableName
+    | interpolation
+    | hexcolor
+    | identifier
+    | expression
+    | block
+    | Lparen Rparen
+    | measurment
+    ;
 
-referenceUrl
-    : StringLiteral
-    | UrlStart Url UrlEnd
+// Function
+functionDeclaration
+    : Function ( namespace_? identifier )?
+    Lparen parameters Rparen BlockStart functionBody? BlockEnd
     ;
 
-asClause
-  : AS (TIMES | identifier)
-  ;
+parameters
+    : parameter? (Comma parameter)*
+    ;
 
-withClause
-  : WITH LPAREN keywordArgument (COMMA keywordArgument)* COMMA? RPAREN
-  ;
+parameter
+    : ( value | variableDeclaration | listSpaceSeparated | mapDeclaration ) arglist? prio?
+    ;
 
-keywordArgument
-  : identifierVariableName COLON expression
-  ;
+functionBody
+    : functionStatement* functionReturn?
```

Comment:
I have the following warning:

```
warning(154): ScssParser.g4:322:0: rule functionDeclaration contains an optional block with at least one alternative that can match an empty string
```

Please fix it by removing optionality from `functionBody`:

```suggestion
    : functionStatement+ functionReturn?
```